pkg/backend/snapshot.go

```diff
@@ -638,6 +639,24 @@ func (sm *SnapshotManager) saveSnapshot() error {
 	return nil
 }
 
+// unsafeServiceLoop doesn't save Snapshots when mutations occur and instead saves Snapshots when
+// SnapshotManager.Close() is invoked. It trades reliability for speed as every mutation does not
+// cause a Snapshot to be serialized to the user's state backend.
+func (sm *SnapshotManager) unsafeServiceLoop(mutationRequests chan mutationRequest, done chan error) {
+	for {
+		select {
+		case request := <-mutationRequests:
+			request.mutator()
+			request.result <- nil
+		case <-sm.cancel:
+			done <- sm.saveSnapshot()
```

Comment:
Hm so it always reaches sm.cancel?

Reply:
```
func (sm *SnapshotManager) Close() error {
	close(sm.cancel)
	return <-sm.done
}

```

I guess this is reasonable. We need to make sure the code that uses SnapshotManager always calls Close - it should. Rain or shine.  Close makes sure cancel chan closes and this logic runs.

Reply:
I was initially considering using defer for extra safety(because of panics), but SnapshotManager has the following engine interface

https://github.com/pulumi/pulumi/blob/a952140b34c667d005116ee53be08077b5863948/pkg/engine/snapshot.go#L24-L25

so it's idiomatic for the caller to make sure to close it and from my checks, it's closed everywhere.

Reply:
Good point about defer, we could throw it in here. `case <-sm.cancel` could return to terminate the loop. But the func itself could say `defer func () { done <- sm.Snapshot() }`. If nothing else it reads better like we really intend this last step to happen.

Reply:
one concern is I've read that closing a closed channel causes a panic.

Reply:
How is this relevant here?

Reply:
ah. I misread your suggestion.